### PR TITLE
[FLINK-9688] [table] ATAN2 sql function support

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

*This PR adds SqlFunction atan2*

## Brief change log

  - *ATAN2 implementation*
  - *tests*
  - *Documentation updates related to Atan2*

## Verifying this change

This change added tests and can be verified as follows:
updates in 
- *SqlExpressionTest.scala*
- *ScalarFunctionsTest.scala*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): ( no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: ( no )
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? ( docs )
in *docs/dev/table/sql.md* and *docs/dev/table/tableApi.md*

### CODE REVIEW

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala [hunk 1]
```diff
@@ -1111,6 +1111,53 @@ class ScalarFunctionsTest extends ScalarTypesTestBase {
...
 
+  @Test
+  def testAtan2(): Unit = {
+    testAllApis(
+      'f25.atan2('f26),
+      "f25.atan2(f26)",
+      "ATAN2(f25, f26)",
+      math.atan2(0.42.toByte, 0.toByte).toString)
+
+
```
Comment (yanghua) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala]: Remove this blank line would be better~

Comment (snuyanzin) [reply flink-libraries/flink-table/src/test/scala/org/apache/flink/table/expressions/ScalarFunctionsTest.scala]: Thank you for that catch. 
blank line removed

File: docs/dev/table/sql.md [hunk 2]
```diff
@@ -1510,6 +1510,17 @@ ATAN(numeric)
       </td>
     </tr>
 
+    <tr>
+      <td>
+        {% highlight text %}
+ATAN2(numeric)
```
Comment (fhueske) [inline docs/dev/table/sql.md]: should be `ATAN2(numeric, numeric)`

File: docs/dev/table/sql.md [hunk 3]
```diff
@@ -1510,6 +1510,17 @@ ATAN(numeric)
...
+{% endhighlight %}
+      </td>
+      <td>
+        <p>Calculates the arc tangent of a given coordinates.</p>
```
Comment (fhueske) [inline docs/dev/table/sql.md]: should be `of a given coordinate.` (-s)?

File: docs/dev/table/tableApi.md [hunk 4]
```diff
@@ -2184,6 +2184,17 @@ NUMERIC.atan()
       </td>
     </tr>
 
+    <tr>
+      <td>
+        {% highlight java %}
+NUMERIC.atan2()
```
Comment (fhueske) [inline docs/dev/table/tableApi.md]: should be `NUMERIC.atan2(NUMERIC)`

File: docs/dev/table/tableApi.md [hunk 5]
```diff
@@ -2184,6 +2184,17 @@ NUMERIC.atan()
...
+{% endhighlight %}
+      </td>
+      <td>
+        <p>Calculates the arc tangent of a given coordinates.</p>
```
Comment (fhueske) [inline docs/dev/table/tableApi.md]: see above

File: docs/dev/table/tableApi.md [hunk 6]
```diff
@@ -3748,6 +3759,17 @@ NUMERIC.atan()
       </td>
     </tr>
 
+    <tr>
+      <td>
+        {% highlight scala %}
+NUMERIC.atan2()
```
Comment (fhueske) [inline docs/dev/table/tableApi.md]: should be `NUMERIC.atan2(NUMERIC)`.

File: docs/dev/table/tableApi.md [hunk 7]
```diff
@@ -3748,6 +3759,17 @@ NUMERIC.atan()
...
+{% endhighlight %}
+      </td>
+      <td>
+        <p>Calculates the tangent of a given coordinates.</p>
```
Comment (fhueske) [inline docs/dev/table/tableApi.md]: see above

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/mathExpressions.scala [hunk 8]
```diff
@@ -243,6 +243,16 @@ case class Atan(child: Expression) extends UnaryExpression {
   }
 }
 
+case class Atan2(left: Expression, right: Expression) extends BinaryExpression {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/mathExpressions.scala]: No input validation? override `expectedTypes` or `validateInput()`

File: docs/dev/table/sql.md [hunk 9]
```diff
@@ -1510,6 +1510,17 @@ ATAN(numeric)
...
     </tr>
 
+    <tr>
+      <td>
+        {% highlight text %}
+ATAN2(numeric, numeric)
+{% endhighlight %}
+      </td>
+      <td>
+        <p>Calculates the arc tangent of a given coordinate.</p>
```
Comment (hequn8128) [inline docs/dev/table/sql.md]: I checked the [wiki](https://en.wikipedia.org/wiki/Atan2) about atan2, it said: The atan2 function calculates one unique arc tangent value from two variables y and x. So, would it be better `of two variables`?  atan2(y,x) can be the arc tangent of (x,y) or (nx, ny).

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala [hunk 10]
```diff
@@ -300,6 +300,24 @@ object FunctionGenerator {
...
+    ATAN2,
+    Seq(DOUBLE_TYPE_INFO, BIG_DEC_TYPE_INFO),
+    DOUBLE_TYPE_INFO,
+    BuiltInMethods.ATAN2_DEC)
+
+  addSqlFunctionMethod(
+    ATAN2,
+    Seq(BIG_DEC_TYPE_INFO, BIG_DEC_TYPE_INFO),
+    DOUBLE_TYPE_INFO,
+    BuiltInMethods.ATAN2_DEC)
```
Comment (hequn8128) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/calls/FunctionGenerator.scala]: BuiltInMethods.ATAN2_DEC_DEC

File: docs/dev/table/sql.md (hunk 3 above)
Comment (hequn8128) [reply docs/dev/table/sql.md]: I checked the [wiki](https://en.wikipedia.org/wiki/Atan2) about atan2, it said: The atan2 function calculates one unique arc tangent value from two variables y and x. So, would it be better `of two variables`?  atan2(y,x) can be the arc tangent of (x,y) or (nx, ny).

File: docs/dev/table/sql.md (hunk 9 above)
Comment (snuyanzin) [reply docs/dev/table/sql.md]: Hello @hequn8128! The idea about coordinates comes from Math.atan2 + Calcite's definition here [1] ATAN2(numeric, numeric) | Returns the arc tangent of the numeric coordinates.
Does it make sense to take the same wording as Calcite has?
[1] https://calcite.apache.org/docs/reference.html

**Comment by snuyanzin** [comment]
@fhueske @hequn8128 @yanghua thank you very much for your review and comments
I did almost all the corrections except one related to atan2 description
So I have a question:
as @hequn8128 wrote
> I checked the [wiki](https://en.wikipedia.org/wiki/Atan2) about atan2, it said: The atan2 function calculates one unique arc tangent value from two variables y and x. So, would it be better of two variables? atan2(y,x) can be the arc tangent of (x,y) or (nx, ny).

At the same time I checked [Calcite's definition](https://calcite.apache.org/docs/reference.html) of it: _ATAN2(numeric, numeric) | Returns the arc tangent of the numeric coordinates._ 
What do you think what is more suitable?

File: docs/dev/table/tableApi.md [hunk 11]
```diff
@@ -2184,6 +2184,17 @@ NUMERIC.atan()
       </td>
     </tr>
 
+    <tr>
+      <td>
+        {% highlight java %}
+NUMERIC.atan2(NUMERIC)
```
Comment (fhueske) [inline docs/dev/table/tableApi.md]: given that both parameters are equally important, we might want to change the syntax to `atan2(Numeric, Numeric)`. IMO, that would be more intuitive. What do you think?

Comment (snuyanzin) [reply docs/dev/table/tableApi.md]: Agree I changed the syntax to `atan2(Numeric, Numeric)`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/mathExpressions.scala [hunk 12]
```diff
@@ -243,6 +243,19 @@ case class Atan(child: Expression) extends UnaryExpression {
   }
 }
 
+case class Atan2(left: Expression, right: Expression) extends BinaryExpression with InputTypeSpec {
```
Comment (twalthr) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/expressions/mathExpressions.scala]: We should not use `InputTypeSpec` here. It forces a certain type which results in BigDecimal even for double types.

**Pull request closed without merging**